### ISSUE TITLE
Archive attachments to azure storage

### ISSUE THREAD

**Issue by venkatkrk** [seanmcne/Microsoft.Xrm.Data.PowerShell.Samples#6]
It will be great to have an example of how we could archive attachments to azure storage in CRM online and store a link in Notes. (we have this program written in C# but could be a great example in powershell)

**Comment by kenakamu** [comment]
This sounds interesting. Do you mind sharing your C# code so that I can re-write to PowerShell?

**Comment by venkatkrk** [comment]
Hi Ken,

Below code is a webjob, which retrieves notes and if the attachment is a image, resizes it and stores it in Azure, then it generates the URL of the image in Azure and stores it back in the note after deleting the original binary image in the note by replacing it with URL. Happy to help if u have questions.

using System;
using System.Collections.Generic;
using System.Configuration;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using System.Text;
using Microsoft.Azure;
using Microsoft.Azure.WebJobs;
using Microsoft.WindowsAzure.Storage;
using Microsoft.Xrm.Client;
using Microsoft.Xrm.Client.Services;
using Microsoft.Xrm.Sdk.Query;
using Newtonsoft.Json;

namespace ResizeImageAndSaveToAzureStorage
{
    public class Program
    {
        static void Main(string[] args)
        {
            var storageConn = ConfigurationManager.ConnectionStrings["AzureStorage"].ConnectionString;

```
        JobHostConfiguration config = new JobHostConfiguration();
        config.StorageConnectionString = storageConn;
        config.DashboardConnectionString = storageConn;
        config.NameResolver = new QueueNameResolver();

        JobHost host = new JobHost(config);
        host.RunAndBlock();
    }

    public static void ProcessQueueMessage([QueueTrigger("%queueName%")] string message, TextWriter logger)
    {
        logger.WriteLine("Got message: " + message);

        var storageAccount = CloudStorageAccount.Parse(ConfigurationManager.ConnectionStrings["AzureStorage"].ConnectionString);
        var blobClient = storageAccount.CreateCloudBlobClient();

        var crmConnection = CrmConnection.Parse(ConfigurationManager.ConnectionStrings["CRMOnline"].ConnectionString);
        var crmService = new OrganizationService(crmConnection);

        var msgObject = JsonConvert.DeserializeObject<QueueMessageJson>(message);
        string blobPath = msgObject.savedFileRelativePath;

        var container = blobClient.GetContainerReference(msgObject.azureStorageContainer);
        var blockBlob1200 = container.GetBlockBlobReference(blobPath);

        var annotation = crmService.Retrieve("annotation", new Guid(msgObject.annotationId), new ColumnSet("notetext"));
        var noteText = annotation.GetAttributeValue<string>("notetext");

        var imagesList = new List<string>();

        using (var stream = new MemoryStream())
        {
            blockBlob1200.DownloadToStream(stream);
            stream.Position = 0;

            var fileExtension = blobPath.Substring(blobPath.LastIndexOf('.') + 1, blobPath.Length - 1 - blobPath.LastIndexOf('.')).ToLower();
            if (fileExtension != "jpg")
            {
                var newBlockBlob1200 = container.GetBlockBlobReference(blobPath.Replace(fileExtension, "jpg"));
                newBlockBlob1200.UploadFromStream(stream);
                stream.Position = 0;

                blockBlob1200.Delete();
                //blockBlob1200 = newBlockBlob1200;
            }

            var image = Image.FromStream(stream);
            using (var newImage = ResizeImage(image, 500, 333))
            {
                using (var ms = new MemoryStream())
                {
                    newImage.Save(ms, ImageFormat.Jpeg);
                    ms.Position = 0;

                    var blockBlob = container.GetBlockBlobReference(blobPath.Replace("1200x800", "500x333").Replace(fileExtension, "jpg"));
                    blockBlob.UploadFromStream(ms);

                    imagesList.Add(blockBlob.Name);
                }
            }
            using (var newImage = ResizeImage(image, 220, 150))
            {
                using (var ms = new MemoryStream())
                {
                    newImage.Save(ms, ImageFormat.Jpeg);
                    ms.Position = 0;

                    var blockBlob = container.GetBlockBlobReference(blobPath.Replace("1200x800", "220x150").Replace(fileExtension, "jpg"));
                    blockBlob.UploadFromStream(ms);

                    imagesList.Add(blockBlob.Name);
                }
            }
        }

        var sb = new StringBuilder();

        if (noteText != null)
        {
            sb.AppendLine(noteText);
        }

        foreach (var image in imagesList)
        {
            sb.AppendLine(string.Format("{0}{1}/{2}", storageAccount.BlobEndpoint.AbsoluteUri, msgObject.azureStorageContainer, image));
        }

        annotation["notetext"] = sb.ToString();
        crmService.Update(annotation);
    }

    private static Bitmap ResizeImage(Image image, int width, int height)
    {
        var destRect = new Rectangle(0, 0, width, height);
        var destImage = new Bitmap(width, height);

        destImage.SetResolution(image.HorizontalResolution, image.VerticalResolution);

        using (var graphics = Graphics.FromImage(destImage))
        {
            graphics.CompositingMode = CompositingMode.SourceCopy;
            graphics.CompositingQuality = CompositingQuality.HighQuality;
            graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;
            graphics.SmoothingMode = SmoothingMode.HighQuality;
            graphics.PixelOffsetMode = PixelOffsetMode.HighQuality;

            using (var wrapMode = new ImageAttributes())
            {
                wrapMode.SetWrapMode(WrapMode.TileFlipXY);
                graphics.DrawImage(image, destRect, 0, 0, image.Width, image.Height, GraphicsUnit.Pixel, wrapMode);
            }
        }

        return destImage;
    }

    public class QueueNameResolver : INameResolver
    {
        public string Resolve(string name)
        {
            return ConfigurationManager.AppSettings[name];
        }
    }
}
```

}

**Comment by DawidPotgieter** [comment]
Hi there,

Not sure if you're still looking for code here, but I've built a crm plugin to transparently store binaries (annotation and activity attachments) in azure here:

https://dynamicscrmbinarystorageoptions.codeplex.com/

I believe the code would be useful in what you're trying to do. I had to use REST calls directly instead of the Azure Storage Client libary, but that's not an issue I believe.

Cheers.
